File: 024-swap-pairs/Solution.java
#迭代寫法，遞迴寫法見python
import java.util.ArrayList;
class ListNode{
    int val;
    ListNode prev;
    ListNode next;
    ListNode(int val, ListNode prev, ListNode next){
        this.val = val;
        this.prev = prev;
        this.next = next;
    }
}
class LinkedList{
    ListNode head;
    ListNode tail;
    LinkedList(ListNode head, ListNode tail){
        this.head = head;
        this.tail = tail;
    }
    void AddAtTail(ListNode newNode){
        if (head == null){
            head = newNode;
            tail = newNode;
            return;
        }
        newNode.prev = tail;
        tail.next = newNode;
        tail = newNode;
    }
    void printList(ListNode head){
        if (head == null){
            System.out.println("head = null");
            return;
        }
        ArrayList<Integer> list = new ArrayList<>();
        list.add(head.val);

        ListNode cur = head;
        while (cur.next != null){
            cur = cur.next;
            list.add(cur.val);
        }
        System.out.println(list);
    }
}
class Solution24 {
    LinkedList mainList;
    Solution24(){
        mainList = new LinkedList(null, null);
    }
    public ListNode swapPairs(ListNode head) {
        if (head == null || head.next == null){
            mainList.printList(head);
            return head;
        }
        ListNode dummy = new ListNode(0, null, head);
        ListNode odd = dummy;
        ListNode even = dummy.next;
        ListNode temp;
        while (even.next != null){
            temp = even.next;
            even.next = even.next.next;
            temp.next = odd.next;
            odd.next = temp;
            //pointers shift
            if (even.next != null){
                even = even.next;
            }
            odd = odd.next.next;
            System.out.println("odd.val = " + odd.val);
            System.out.println("even.val = " + even.val);
            mainList.printList(dummy);
        }
        return dummy.next;
    }

    static void main(String[] args) {
        Solution24 sol = new Solution24();
        int[] nums = {1, 2, 3, 4};
        ListNode newNode;
        for (int i: nums){
            newNode = new ListNode(i, null, null);
            sol.mainList.AddAtTail(newNode);
        }
        sol.swapPairs(sol.mainList.head);
    }
}
/*
Loop1:
odd        even    temp
  ↓        ↓        ↓
[d|-]--->[1|-]--->[2|-]--->[3|-]--->[4|-]

temp--->[2|-]
odd         even
  ↓         ↓
[d|-]--->[1|-]--->[3|-]--->[4|-]

odd        temp    even
  ↓         ↓       ↓
[d|-]--->[2|-]--->[1|-]--->[3|-]--->[4|-]

//pointers shift

                 odd        even    temp
                  ↓         ↓        ↓
[d|-]--->[2|-]--->[1|-]--->[3|-]--->[4|-]

Loop2:
                 odd        even    temp
                  ↓         ↓        ↓
[d|-]--->[2|-]--->[1|-]--->[3|-]--->[4|-]

temp--->[4|-]
                 odd
                  ↓
[d|-]--->[2|-]--->[1|-]--->[3|-]

                 odd        even
                  ↓         ↓
[d|-]--->[2|-]--->[1|-]--->[4|-]--->[3|-]

//pointers shift

                                     odd        even
                                      ↓         ↓
[d|-]--->[2|-]--->[1|-]--->[4|-]--->[3|-]--->null

exit
 */

File: 024-swap-pairs/solution.py
#遞迴寫法，迭代寫法見java
from typing import Optional
class ListNode:
    def __init__(self, val, next) -> None:
        self.val = val
        self.next = next
class LinkedList:
    def __init__(self) -> None:
        self.head = None
    def AddAtTail(self, newNode):
        if self.head is None:
            self.head = newNode
            return
        cur = self.head
        while cur.next is not None:
            cur = cur.next
        cur.next = newNode
    def printList(self, head):
        if self.head is None:
            return head
        nums = []
        nums.append(head.val)
        cur = head
        while cur.next is not None:
            cur = cur.next
            nums.append(cur.val)
        print(nums)
        
class Solution1122:
    def __init__(self) -> None:
        self.layer = 1
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        print("layer = ", self.layer)#debug

        if not head or not head.next:
            print("head or head.next is None, break")
            return head
        newHead = head.next
        self.layer += 1#debug
        print("head = ", head.val)#debug
        print("newHead = ", newHead.val)#debug
        head.next = self.swapPairs(newHead.next)
        newHead.next = head
        return newHead
"""
初始化:
head
  |
  V 
[1] -> [2] -> [3] -> [4] -> None

第一層swapPairs
newHead = head.next
head   newHead
  |      |
  V      V
[1] -> [2] -> [3] -> [4] -> None

head.next = self.swapPairs(newHead.next)
head   newHead
  |      |
  V      V
[1] -> [2] -> [3] -> [4] -> None
                              ^
                              |
                              等待 swapPairs(ListNode(3)) 的返回值
第二層swapPairs:
newHead = head.next
(遞迴中的 head) (遞迴中的 newHead)
           |                  |
           V                  V
         [3] ->             [4] -> None
        
head.next = self.swapPairs(newHead.next)
(遞迴中的 head) (遞迴中的 newHead)
           |                  |
           V                  V
         [3] ->             [4] -> None
                                    ^
                                    |
                                    等待 swapPairs(None) 的返回值
第三層swapPairs:
if not head or not head.next:
    return head
    
返回第二層swapPairs:
head.next = self.swapPairs(newHead.next)
(遞迴中的 head) (遞迴中的 newHead)
           |                  |
           V                  V
         [3] -> None        [4] -> None  (注意：此時 4 的 next 仍然指向 None，它沒有被改變)
         
head.next = None
(遞迴中的 head) (遞迴中的 newHead)
           |                  |
           V                  V
         [3] <-             [4] -> None
           ^                  |
           |                  |
           -------------------

newHead.next = head
(遞迴中的 head) (遞迴中的 newHead)
           |                  |
           V                  V
         [3] <-             [4] -> None
           ^                  |
           |                  |
           -------------------
           
現在 4 -> 3 -> None 這段鏈結串列已經形成。  

return newHead
返回第一層swapPairs:
返回 newHead，也就是 ListNode(4)。


返回第一層swapPairs:
head.next = self.swapPairs(newHead.next)
(原始 head) (原始 newHead)
         |             |
         V             V
       [1] ->        [2]
         ^             |
         |             |
         ---------------------------------> [4] -> [3] -> None
         
head.next = ListNode(4)
newHead.next = head
(原始 head) (原始 newHead)
         |             |
         V             V
       [1] <-        [2]
         ^             |
         |             |
         ---------------------------------> [4] -> [3] -> None
         ^                                  |
         |----------------------------------
現在 2 -> 1 -> 4 -> 3 -> None 這段鏈結串列已經形成

return newHead  
返回 newHead，也就是 ListNode(2)。
"""

if __name__ == "__main__":
    nums = [1, 2, 3, 4]
    sol = Solution1122()
    mainList = LinkedList()
    for i in nums:
        newNode = ListNode(i, None)
        mainList.AddAtTail(newNode)
    sol.swapPairs(mainList.head)